[Lee Xiong Jie, Isaac] iP

luminousleek opened the pull request:
# Bob
> Bob the Builder, can we fix it?
> Bob, the Builder, yes we can! - [some children's TV show apparently](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Bob doesn't build though, he only stores tasks. Here's some tasks he can store:
* Items to do
* Deadlines
* Events

Here's how you can use Bob:
1. Download the jar file
2. Double-click it
3. ???
4. ~~PROFIT!~~ 

Stuff Bob can do:
- [x] Add and remove tasks
- [x] Find tasks that match a keyword
- [x] Store tasks as a txt file on your device
- [ ] Have a GUI (eventually)
- [ ] Build 🏗️ 
- [ ] Fix it 👨‍🔧 

Unfortunately, Bob cannot print `Hello World!`. But you can with this code snippet!
```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World!");
    }
}
```

Commit:
Add Gradle support

luminousleek committed:
Implement Greet, Echo and Exit

luminousleek committed:
Implement ability to store text and display stored text in an ordered list

luminousleek committed:
Add ability to mark tasks as done. Also added a Task class

luminousleek committed:
add Deadline, Event and ToDo classes

luminousleek committed:
fix bugs in Deadline and Event

luminousleek committed:
Add support for ToDo, Deadline and Event

luminousleek committed:
Add test cases for Duke

luminousleek committed:
added DukeException and error handling, add more test cases

luminousleek committed:
change Task array to ArrayList, fix bug for validating Event input

luminousleek committed:
fix bug when list did not have newline characters

luminousleek committed:
added ability to remove tasks, more testcases

luminousleek committed:
Make Task an abstract class with abstract printToFile method

luminousleek committed:
Implement printToFile method in subclasses

luminousleek committed:
Remove commented out Duke logo

luminousleek committed:
Add overloaded constructors for Deadline and Event

luminousleek committed:
Add ability to save and load task list from a local file

luminousleek committed:
Update tests to take saving and loading into account

luminousleek committed:
Update .gitignore to ignore taskList.txt files

luminousleek committed:
Remove commented out logo

luminousleek committed:
Add ability to store Dates in Events and Deadlines

luminousleek committed:
Merge branch 'branch-Level-7'

luminousleek committed:
Resolve merge conflict with branch-Level-8

luminousleek committed:
Update Deadline and Event to work with saving to file

luminousleek committed:
Update tests to test Date handling

luminousleek committed:
Update .gitignore to include testing taskList.txt

luminousleek committed:
Add Parser, Storage, TaskList and Ui classes

luminousleek committed:
Add Command class and subclasses

luminousleek committed:
Add TaskTypes enum

luminousleek committed:
Modify Duke to make use of new classes

luminousleek committed:
Update tests

luminousleek committed:
Organise classes into packages

luminousleek committed:
Add JUnit tests for Event and Parser classes

luminousleek committed:
Add manifest for JAR file

luminousleek committed:
Add JavaDocs to all the classes

luminousleek committed:
Add checkstyle and suppressions file for style checking

luminousleek committed:
Edit files to match coding standard

luminousleek committed:
Add ability to find tasks containing a given keyword

luminousleek committed:
Merge branch 'branch-A-JavaDoc'

luminousleek committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

luminousleek committed:
Merge branch 'branch-Level-9'

luminousleek committed:
Make findTask use the Task.getName() method

luminousleek committed:
Add missing JavaDocs and fix style

luminousleek committed:
Merge remote-tracking branch 'origin/add-gradle-support'

luminousleek committed:
Add missing JavaDoc so Gradle can build

luminousleek committed:
Change order of import statements so Gradle can build

luminousleek committed:
Merge branch 'branch-A-Gradle'

luminousleek committed:
Add JavaFX dependencies to build.gradle

luminousleek committed:
Add GUI using JavaFX

luminousleek committed:
Merge branch 'branch-Level-10'

syoopie commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,72 @@
+        return generateMsg(newTask, tasks);
+    }
+
+    private String generateMsg(Task tsk, TaskList tasks) {
+        String msg = "Got it. I've added this task:\n  " + tsk + "\nNow you have " + tasks.numTasks();
+        msg = tasks.numTasks() == 1 ? msg + " task in the list" : msg + " tasks in the list.";
+        return msg;
+    }
```
Maybe the argument `tsk` can be renamed to `task` for more clarity. But honestly I am commenting on this because there is nothing else to comment on... 🤡

syoopie commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,147 @@
+     * @param num the index corresponding to the desired Task.
+     * @return the Task corresponding to the given index.
+     */
+    public Task getTask(int num) {
```
Perhaps `num` can just be renamed to `index` for more clarity.

syoopie reviewed (commented):
Amazing that you have already moved on to Gradle and GUI :D I still feel so far behind

deetoomoo commented on src/main/java/Launcher.java:
```diff
@@ -0,0 +1,10 @@
+import javafx.application.Application;
```
Perhaps you should group this class in a package? Might help reviews understand the code better!

deetoomoo reviewed (commented):
Overall good adherence to the Java coding standard! Good job!

nhjryan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+    public static Command parse(String str) throws DukeException {
+        String[] strArr = str.split("\\s", 2);
+
+        switch (strArr[0]) {
```
I really like the use of switch statements to handle exceptional inputs and re-direct the input to the necessary command methods, really helps make the code look neat and understandable.

nhjryan commented on src/main/java/duke/command/RestoreCommand.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * The RestoreCommand is given when the user wants to revert to a previously saved TaskList.
+ */
+public class RestoreCommand extends Command {
```
I like how you took into account the situation where the list is accidentally deleted. :)

nhjryan reviewed (commented):
LGTM! All the code that I've read adheres to the specified coding standard and is laid out in a neat and readable manner.

luminousleek committed:
Add assertion to AddCommand

AddCommand has no way to verify if a Task will be added to the TaskList.

We want to ensure that a Task will be added to the TaskList instead of a
null value.

Let's add an assert statement to verify the assumption that a Task will be
added.

This is done with an assertion statement so I can fulfill the
requirement for week 5 :/

Thanks for your patience and understanding.

luminousleek committed:
Remove dead code for text-based UI

The files still contained code for the old text-based UI.

This code is now redundant with the GUI, and causes the code to be
harder to read.

Let's remove this redundant code.

luminousleek committed:
Remove magic literals for Event and Deadline and abstract formatDate

The keywords for Event and Deadline were literals, and the code
to format the date for printing was duplicated.

This is bad for changing the keywords for Event and Deadline in
the future. It's also bad for changing the date format for printing
as someone might forget to change the format in both classes.

Let's
* Make the keywords a constant class field of the respective classes
* Abstract out the function to format the date
* Make the date format a constant of Task

luminousleek committed:
Add Exception subclasses

There used to be only one DukeException class to encapsulate all
the errors in Duke.

This made throwing a new error tedious as a new error message had
to be passed in every time, and editing error messages would also
be a pain.

Let's
* Create new exceptions for common exceptions
* Move all these new exceptions to the duke.exception package

luminousleek committed:
Implement new DukeException subclasses

The previous commit did not contain the constructors and details
for the new DukeEception subclasses.

This means they wouldn't work.

Let's actually make them work.

luminousleek committed:
Make Parser follow SLAP

The parse method in Parser was quite clunky with multiple layers of
abstraction.

This made it hard to read and understand, and also had duplicated
code segments.

Let's
* Create new methods for parsing each command in the similar style
as prepareAdd
* Abstract out the code to check string array lengths as functions
* Use the new DukeException subclasses
* Amend the import statements accordingly

luminousleek committed:
Create a constant COMMAND_WORD class level field

Previously the command words for Tasks were literals.

This made changing the command words difficult.

Let's make class level COMMAND_WORD fields in each Task subclass.

luminousleek committed:
Use new DukeException subclasses

Previously only the parent DukeException class was used.

This meant that new error messages had to be used for different
types of errors.

Let's
* Implement the new DukeException subclasses
* Amend the import statements accordingly

luminousleek committed:
Remove unary constructor for Event and Deadline

The current unary constructors for Events and Deadlines use the
String.substring() method so that it can call the constructor of
their superclass.

This is pretty clunky because the same thing can be done with the
String.split() method which is a lot more understandable.

Let's
* Add methods in AddCommand to split the input strings for Deadlines
and Events
* Shift the method to check whether the keyword exists from Parser
to AddCommand
* Remove the unary constructor
* Update EventTest with the new constructor

luminousleek committed:
Remove Ui from Command

luminousleek committed:
Remove Ui from JavaDocs

luminousleek committed:
Refactor GUI code

The current code to display the JavaFX Nodes for the GUI is mostly
in the Duke and Ui classes.

This makes it very hard to read, and also difficult to adjust the
various UI components.

Let's
* Abstract out all the components as new subclasses
* Rewrite the Ui class to encapsulate everything for the GUI
* Shift the handleInput method to the Ui class
* Put all the UI components into the ui package
* Simplify the Duke class massively

This took way too long. I'm done with this branch.

luminousleek committed:
Merge pull request #1 from luminousleek/branch-A-Assertions

Add assertion to AddCommand

luminousleek committed:
Merge pull request #2 from luminousleek/branch-A-CodeQuality

Improve Code Quality

luminousleek committed:
Add TimeTask abstract class

Deadlines and Events are subclasses of TimeTask

luminousleek committed:
Add function to edit names and times of tasks

luminousleek committed:
Add space between numbers and tasks in listTaskArr()

luminousleek committed:
Merge branch 'branch-C-Update'

luminousleek committed:
Make Find case-insensitive

luminousleek committed:
Add User Guide and Product Screenshot

luminousleek committed:
Shift resources folder to be under main folder

luminousleek committed:
Fix bug where data directory was not created in first startup

luminousleek committed:
Update Manifest

luminousleek committed:
Fix CheckStyle errors

luminousleek committed:
Update build.gradle's main class to Launcher

luminousleek committed:
Fix bug where images did not appear in jar
